brmp/numpyro_backend.py

```diff
@@ -39,15 +40,22 @@ def from_numpy(data):
 
 # TODO: Better name.
 def run_model_on_samples_and_data(modelfn, samples, data):
-    return vmap(lambda sample: handler.substitute(modelfn, sample)(**data, mode='prior_and_mu'))(samples)
+    assert type(samples) == dict
+    assert len(samples) > 0
+    num_chains, num_samples = next(iter(samples.values())).shape[0:2]
+    assert all(arr.shape[0:2] == (num_chains, num_samples) for arr in samples.values())
+    flat_samples = {k: flatten(arr) for k, arr in samples.items()}
+    out = vmap(lambda sample: handler.substitute(modelfn, sample)(**data, mode='prior_and_mu'))(flat_samples)
```

Comment:
What do the the different modes do - in particular, `prior_and_mu`, `prior` and `full`? It seems to me that this will generate samples from the posterior predictive, but I am not sure if I fully follow the distinction between the three modes. I think that you would also want to `vmap` over the rng seeds since this will use the same random stream for each row. NumPyro will have an [infer_util.predictive](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/infer_util.py#L293) function in the next release that you may find useful. cc. @fehiepsi

Reply:
I think that we just need to use @neerajprad's trick: `numpyro.sample("mu", dist.Delta(mu), obs=mu)` and `predictive` utility for these purposes. Yesterday I tried to port chapters 4 & 5 of Statistical Rethinking book to see limitations of NumPyro API and found that that trick is pretty handy. I think what you want to achieve is something similar to code blocks 5.3, 5.4, 5.5 in [this gist](https://nbviewer.jupyter.org/gist/fehiepsi/195e080f7513f34402fefa82b16e7248#Code-5.3)?

Reply:
> What do the the different modes do

The `mode` flag makes it possible to skip parts of the generated model when it is run. Something like:

||`prior_only`|`prior_and_mu`|`full`|
|---|---|---|---|
|Sample latents?| Y | Y | Y |
|Compute `mu`?| N | Y | Y |
|Observe `y`?| N | N | Y |

(Where `mu` is the value computed for the location of the response distribution before applying any inv. link fn.)

In case the question was more about where this is used...

The `prior_only` mode is used when running the model in order to collect samples from the prior, for example.

`prior_and_mu` is used to e.g. implement the [`location`](https://github.com/pyro-ppl/brmp/blob/11f224f693ad55851429d914178a85c369a35e2e/brmp/backend.py#L32-L40) method required of a back end. (Which is used for e.g. sampling the posterior predictive as you mentioned.) Internally, this makes use of `run_model_on_samples_and_data`, which replays a model on a bag of samples at the given data set in order to compute a `mu` vector. The `prior_and_mu` flag avoids us having to provide a bunch of observations to the model in order to do so.

(It's worth mentioning the pyro back end has a similar `mode` flag, which finds additional uses in the implementation of the `svi` method.)

Does that help? I'm happy to try to explain more carefully if not.

>  I think that you would also want to vmap over the rng seeds since this will use the same random stream for each row.

`run_model_on_samples_and_data` itself is deterministic. (Taking samples and the data as input, and computing only `mu`.) Sampling from the posterior predictive happens elsewhere. More specifically, it happens in the implementation of the [`sample_response_fn`](https://github.com/pyro-ppl/brmp/blob/11f224f693ad55851429d914178a85c369a35e2e/brmp/backend.py#L86-L89) required of a back end. For the numpyro backend, this is seeded [here](https://github.com/pyro-ppl/brmp/blob/11f224f693ad55851429d914178a85c369a35e2e/brmp/numpyro_codegen.py#L363).

Does that makes sense?

Reply:
I see, that makes sense, and we do not need any seeding to happen in that case. Thanks for explaining it so well! I will take another look at how the generated code looks like for the three modes that you mentioned. I think @fehiepsi's suggestion might be applicable there, but in any case, that's more of an internal detail and it doesn't matter how we record the value of `mu`.